Add sock handler to support graphQL subscriptions

rogelio-o opened the pull request:
Fixes #1180

This is a proposal to support graphQL subscriptions adding a socketHandler (GraphQLSocketHandler) to a SockJSHandler.

There is an example here: [https://github.com/rogelio-o/graphql-vertx-example/blob/master/src/main/java/com/rogelioorts/training/graphql/GraphQLVerticle.java](https://github.com/rogelio-o/graphql-vertx-example/blob/master/src/main/java/com/rogelioorts/training/graphql/GraphQLVerticle.java)

```
final SockJSHandler sockJSHandler = SockJSHandler.create(vertx, sockOptions);
sockJSHandler.socketHandler(GraphQLSocketHandler.create(graphQL));
```

If you are all ok with the proposal, I can work in add documentation and test in the same PR for this new feature.

rogelio-o commented:
What do you think @pmlopes @tsegismont?

tsegismont commented:
@rogelio-o thanks for sharing!

So the assumption is users would send only subscription requests over the socket. However they also want queries and mutations. To support this, some sort of protocol on top of the socket would be needed.

Also, how about implementing this on top of the event bus instead of sockjs? Then GraphQL could be used inside a Vert.x cluster as well as all event bus bridges (including the sockjs one).

rogelio-o commented:
@tsegismont thank you for the feedback!

About the query/mutations request over the socket: totally agree! I'm going to implement it like in Apollo's subscriptions-transport-ws.

About to implement it on top of event bus: How do you propose to implemente it? With Event Bus we have messages but no bidirectional connection between client and server, so we cannot write back the subscriptions on the client connection. What I can do is to implement a Fetcher that listens events directly from event bus and write it on the Observable. I think it is a good solution, the ways to configure GraphQL still being one: through fetchers.

tsegismont commented:
Le mar. 2 juil. 2019 à 21:05, Rogelio Orts <notifications@github.com> a
écrit :

> @tsegismont <https://github.com/tsegismont> thank you for the feedback!
>
> About the query/mutations request over the socket: totally agree! I'm
> going to implement it like in Apollo's subscriptions-transport-ws.
>

FYI, I'm currently working on supporting subscriptions-transport-ws.

If you do something with sockjs, perhaps it would be good to reach out to
Apollo developers. I know they explored this option at some point, but
abandoned it.

> About to implement it on top of event bus: How do you propose to
> implemente it? With Event Bus we have messages but no bidirectional
> connection between client and server, so we cannot write back the
> subscriptions on the client connection. What I can do is to implement a
> Fetcher that listens events directly from event bus and write it on the
> Observable. I think it is a good solution, the ways to configure GraphQL
> still being one: through fetchers.
>

I don't have any proposal. I wanted to know if you had thought about it.

rogelio-o commented:
I think that with my last commit it is compatible with Apollo clients and "subscriptions-transport-ws". I'll update my example to try.

I'm gonna keep working on it and I'll propose you a way to work with Event Bus.

Thanks @tsegismont!

rogelio-o commented:
I think the best way to integrate Event Bus is with a custom fetcher, so the way we configure GraphQL still being the same as for query and mutations, two different ways to configure it can be confusing: https://github.com/vert-x3/vertx-web/pull/1325/commits/40d16eb6176b9ab70b3019dee5da41792b20032d

With this other two commits, raw websocket is used instead of sockjs and different types of messages are added, so it should be compatible with Apollo client: https://github.com/vert-x3/vertx-web/pull/1325/commits/1ac64d15871f455c1a00b31e755491da08f88484 and https://github.com/vert-x3/vertx-web/pull/1325/commits/b0db9d501bb306f92b29da85d95c813bc2d51f09

I'm going to upload a full example and still working on support queries and mutations through the websocket.

tsegismont commented:
> I think that with my last commit it is compatible with Apollo clients and "subscriptions-transport-ws". I'll update my example to try.

Good news. So given you're making good progress, I'll stop working on the Apollo WS transport support. Instead, I'll focus on adding automated tests with Apollo to the build.

>I think the best way to integrate Event Bus is with a custom fetcher, so the way we configure GraphQL still being the same as for query and mutations, two different ways to configure it can be confusing: 40d16eb

Given the progress you're making on Apollo WS transport, it would be best to stop thinking about that for now and put all efforts on the implementation/tests/documentation.

> I'm going to upload a full example and still working on support queries and mutations through the websocket.

Great!

tsegismont commented:
@rogelio-o I will review any updates you can make this week, but then I'm away from the office for 3 weeks. So don't be surprised if you don't get feedback for a while.

tsegismont commented:
@rogelio-o I would like this to go in 3.8.1 so, when I'm back, I'll ping you to see how we can move forward. In the worst case I'll branch off of your fork and add some commits.

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/GraphQLSocketHandler.java:
```diff
@@ -0,0 +1,16 @@
+import io.vertx.core.http.ServerWebSocket;
+import io.vertx.ext.web.handler.graphql.impl.GraphQLSocketHandlerImpl;
+
+public interface GraphQLSocketHandler extends Handler<ServerWebSocket> {
```
For consistency with the other `vertx-web` utilities, it would be better to extend `Handler<RoutingContext>` and then implement the websocket handshake.

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/GraphQLSocketHandler.java:
```diff
@@ -0,0 +1,16 @@
+
+public interface GraphQLSocketHandler extends Handler<ServerWebSocket> {
+
+  GraphQLSocketHandler endHandler(Handler<ServerWebSocket> endHandler);
```
What usages do you have in mind for this callback?

A callback for handshake would be useful since, for example, users could extract an auth token from headers

rogelio-o replied:
The callback is called when a websocket connection is closed. Maybe the Handler can evolve and callbacks can be added when they are asked for. So, I can delete the endHandler.

Due to the handshake is an HTTP call, I think manipulating the HTTP request should be done adding handlers to the path in a previous order. For regular endpoints it is done in that way and having a unique way to do things is a good practice.

tsegismont replied:
>The callback is called when a websocket connection is closed. Maybe the Handler can evolve and callbacks can be added when they are asked for. So, I can delete the endHandler.

Ok, let's start without it.

>Due to the handshake is an HTTP call, I think manipulating the HTTP request should be done adding handlers to the path in a previous order. For regular endpoints it is done in that way and having a unique way to do things is a good practice.

Fair enough. We can revisit later if it proves to be inconvenient. Make sure to document how to authenticate a pure WS client though.

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/GraphQLSocketHandler.java:
```diff
@@ -0,0 +1,16 @@
+
+  GraphQLSocketHandler endHandler(Handler<ServerWebSocket> endHandler);
+
+  static GraphQLSocketHandler create(GraphQL graphQL) {
```
Users may want to send queries and mutations over WS but no subscriptions. How would we handle this case? How about `ApolloWSHandlerOptions` with `subscriptionEnabled` field?

rogelio-o replied:
Does it make sense have this option? Maybe just not defining suscriptions in the GraphQL specification would be enough.

tsegismont replied:
Fair enough.

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/VertxEventBusFetcher.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class VertxEventBusFetcher<T> implements DataFetcher<CompletionStage<Publisher<T>>> {
```
Let's move this out this PR's scope

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLMessage.java:
```diff
@@ -0,0 +1,52 @@
+package io.vertx.ext.web.handler.graphql.impl;
```
Don't forget license headers on all new files.

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLMessage.java:
```diff
@@ -0,0 +1,52 @@
+
+import com.fasterxml.jackson.annotation.JsonValue;
+
+public class GraphQLMessage {
```
Don't forget documentation on all new files

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLMessage.java:
```diff
@@ -0,0 +1,52 @@
+
+  private Type type;
+
+  public String getId() {
```
Can you rename the Java field to `opId` for clarity and use Jackson annotations to keep the JSON `id` field name?

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLSocketHandlerImpl.java:
```diff
@@ -0,0 +1,179 @@
+    }
+  }
+
+  private void sendMessage(ServerWebSocket serverWebSocket, String opId, GraphQLMessage.Type type, ExecutionResult payload) {
```
How about getting rid of the `GraphQLMessage` hierarchy and use `JsonObject`? It seems working with types does not simplify the code.

tsegismont commented:
@rogelio-o when #1328 is merged, can you rebase your PR on master and add a ws-link subscription test? Thanks!

rogelio-o commented:
> @rogelio-o I will review any updates you can make this week, but then I'm away from the office for 3 weeks. So don't be surprised if you don't get feedback for a while.

No problem! Thanks for letting me know.

> @rogelio-o here's a first batch of comments / concerns. Keep up the great work !

Thanks for the comments. I’ll implement them this weekend and keep working on tests and documentation.

> @rogelio-o when #1328 is merged, can you rebase your PR on master and add a ws-link subscription test? Thanks!

Sure!

I’ll keep pushing and commenting on this PR, so you can see the progress when you come back. Enjoy your holidays!

tsegismont commented:
Thank you @rogelio-o !

rogelio-o commented:
Added support for queries and mutations over the websocket.

There is a "problem" related with the WebsocketSubprotocol: to use the ApolloWSHandler, the wsSubprotocol "graphql-ws" has to be added to HttpServerOptions. @tsegismont, do you know an alternative to avoid needing configure the HttpServer? You can see an example of this in ApolloTestsServer.

rogelio-o commented:
Tests and documentation added.

tsegismont commented:
>There is a "problem" related with the WebsocketSubprotocol: to use the ApolloWSHandler, the wsSubprotocol "graphql-ws" has to be added to HttpServerOptions. @tsegismont, do you know an alternative to avoid needing configure the HttpServer? You can see an example of this in ApolloTestsServer.

I looked into the websocket handshake code and you are right, the `HttpServerOptions` is the only source of truth. Ideally, when handshake is handled manually (like when you do auth over websocket for example), users should be able to provide a subprotocol.

One might want different subprotocols to be supported in the handshake on different endpoints. @vietj should we file an issue about this in vertx core?

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/ApolloWSHandlerImpl.java:
```diff
@@ -0,0 +1,243 @@
+
+      @Override
+      public void onError(Throwable t) {
+        sendError(serverWebSocket, opId, t);
```
Shouldn't the subscription be removed here?

rogelio-o replied:
You are right. In Observable documentation, talking about onError: "It will not make further calls to onNext or onCompleted"

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/ApolloWSHandlerImpl.java:
```diff
@@ -0,0 +1,243 @@
+    String opId = message.getString("id");
+
+    // Unsubscribe if it's subscribed
+    if (subscriptions.containsKey(opId)) {
```
How could this happen?

rogelio-o replied:
By sending a start message with an opId and then sending again the a start message with the same opId. Maybe by error or whatever, but it'd receive duplicate data if we don't unsubscribe it before.

tsegismont replied:
Ok, thanks for the details. So that would be if the Apollo client gets buggy or a user writes a custom client.

rogelio-o commented:
> Great job @rogelio-o ! Thank you. Some concerns to address

Thanks! I'll take a peek of the changes in the following days.

rogelio-o commented:
Done!

tsegismont commented:
@rogelio-o I thought again about use cases and it occurred to me that we're missing a couple of things:

- [customization of the context](https://vertx.io/docs/vertx-web-graphql/java/#_providing_data_fetchers_with_some_context) object
- [data loader registry](https://vertx.io/docs/vertx-web-graphql/java/#_using_dataloaders) creation

Do you have some extra time to propose something in this PR? Otherwise I can work on it after it's merged.

tsegismont commented:
@rogelio-o can you please sign the Eclipse Contributor Agreement ?

rogelio-o commented:
> @rogelio-o I thought again about use cases and it occurred to me that we're missing a couple of things:
> 
> * [customization of the context](https://vertx.io/docs/vertx-web-graphql/java/#_providing_data_fetchers_with_some_context) object
> * [data loader registry](https://vertx.io/docs/vertx-web-graphql/java/#_using_dataloaders) creation
> 
> Do you have some extra time to propose something in this PR? Otherwise I can work on it after it's merged.

Sure! I’m out this week, but I could work on it from next Wednesday.

rogelio-o commented:
> @rogelio-o can you please sign the Eclipse Contributor Agreement ?

Done!

tsegismont commented:
>Sure! I’m out this week, but I could work on it from next Wednesday.

Great!

tsegismont commented:
Fixes #1180

tsegismont commented:
@rogelio-o as 3.8.1 might happen sooner than we thought, can you confirm you can take care of the remaining changes?

rogelio-o commented:
> @rogelio-o as 3.8.1 might happen sooner than we thought, can you confirm you can take care of the remaining changes?

Do you mean "customization of the context object" and "data loader registry creation"? Sure! 

"Customization of the context object" done. Tomorrow morning I'll work on the "data loader registry creation".

zenios commented:
I went over the PR and one small thing is missing in order to be fully compliant with Apollo subscription protocol.

The missing piece is the keep alive messages and its easy to implemented.What is really needed is once the connection is established, if keep alive is enabled we reply first with connection_ack and then keep alive (ka). After that we periodically send ka messages.

rogelio-o commented:
Thanks for the feedback @zenios. I've added the feature.

Please @tsegismont, can you review the last commits? Thanks!

zenios commented:
@rogelio-o  What are your comments on implementing something like this https://www.apollographql.com/docs/graphql-subscriptions/lifecycle-events/ on top of your implementation?

I really like how sockjs eventbus bridge gives you the option to subscribe to events such as SOCKET_CREATED, SOCKET_CLOSED, REGISTER, UNREGISTER and i think it would be good to have on the subscriptions implementation as well.

Especially since there is no way (From what is my understanding) to know when someone has cancelled the subscription from within the publisher.

rogelio-o commented:
> @rogelio-o What are your comments on implementing something like this https://www.apollographql.com/docs/graphql-subscriptions/lifecycle-events/ on top of your implementation?
> 
> I really like how sockjs eventbus bridge gives you the option to subscribe to events such as SOCKET_CREATED, SOCKET_CLOSED, REGISTER, UNREGISTER and i think it would be good to have on the subscriptions implementation as well.
> 
> Especially since there is no way (From what is my understanding) to know when someone has cancelled the subscription from within the publisher.

We decided to let out of the scope of this PR the implementation of the handlers, but I think it’s a good idea to take the lifecycle defined by Apollo as guide to a future implementation.

Regarding to the implementation on top of sockJS, it was the first implementation but we decided to do it in the current way because it’s more flexible (the client isn’t forced to use sockJS). What you mention could be achieved adding handlers for “start” and “stop” operations.

zenios commented:
Yes but there are no start stop handler at the moment.

As for connected disconnected i can use socket handshake for connected and
end handler for disconnected.but on the end handler i dont have access to
the  original routing context (think of user management and a list of
connected users)

Also start can be sort of achieved through context handler so what is
actually missing is the stop handler

As a user i would prefer a single point to  handle connected, disconnected,
start, stop instead of jumping between different handlers in order to
somewhat achieve the same effect.

On Sat, 17 Aug 2019, 18:40 Rogelio Orts, <notifications@github.com> wrote:

> @rogelio-o <https://github.com/rogelio-o> What are your comments on
> implementing something like this
> https://www.apollographql.com/docs/graphql-subscriptions/lifecycle-events/
> on top of your implementation?
>
> I really like how sockjs eventbus bridge gives you the option to subscribe
> to events such as SOCKET_CREATED, SOCKET_CLOSED, REGISTER, UNREGISTER and i
> think it would be good to have on the subscriptions implementation as well.
>
> Especially since there is no way (From what is my understanding) to know
> when someone has cancelled the subscription from within the publisher.
>
> We decided to let out of the scope of this PR the implementation of the
> handlers, but I think it’s a good idea to take the lifecycle defined by
> Apollo as guide to a future implementation.
>
> Regarding to the implementation on top of sockJS, it was the first
> implementation but we decided to do it in the current way because it’s more
> flexible (the client isn’t forced to use sockJS). What you mention could be
> achieved adding handlers for “start” and “stop” operations.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-web/pull/1325?email_source=notifications&email_token=AAAV3YIKBPVJY4RXPXFNKFDQFALWDA5CNFSM4H4MH5O2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4QN7AI#issuecomment-522248065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAV3YP7G77EKWR7KVVWV6LQFALWDANCNFSM4H4MH5OQ>
> .
>

rogelio-o commented:
> Yes but there are no start stop handler at the moment.

We should work on the handlers, but coupling the solution to the sockJS library is less flexible approach, IMHO.

> As a user i would prefer a single point to handle connected, disconnected, start, stop instead of jumping between different handlers in order to somewhat achieve the same effect.

Can be done with the approach of a handler per event. In a lambda function, you can call a handler with two arguments in which the first argument could be the kind of event. Also, you could use event bus to send events on each handler and you would have the solution you proposed previously with the bridge.

I think you are right about handlers, they are important. Although they were out of the scope, I'm going to try adding them. What do you think @tsegismont? Are we in time?

zenios commented:
On Sun, 18 Aug 2019, 13:30 Rogelio Orts, <notifications@github.com> wrote:

> Yes but there are no start stop handler at the moment.
>
> We should work on the handlers, but coupling the solution to the sockJS
> library is less flexible approach, IMHO.
>

I am not saying it should be coupled with the sockjs implementation. It was
just a reference on how adding a handler on the sockjs eventbus you could
get events of connected disconnected subscribed and unsubscribed (
https://vertx.io/docs/vertx-web/java/#_handling_event_bus_bridge_events)

> As a user i would prefer a single point to handle connected, disconnected,
> start, stop instead of jumping between different handlers in order to
> somewhat achieve the same effect.
>
> Can be done with the approach of a handler per event. In a lambda
> function, you can call a handler with two arguments in which the first
> argument could be the kind of event. Also, you could use event bus to send
> events on each handler and you would have the solution you proposed
> previously with the bridge.
>
> I think you are right about handlers, they are important. Although they
> were out of the scope, I'm going to try adding them. What do you think
> @tsegismont <https://github.com/tsegismont>? Are we in time?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-web/pull/1325?email_source=notifications&email_token=AAAV3YKELH7KF6L2AJ4JX3LQFEQF5A5CNFSM4H4MH5O2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4Q5APY#issuecomment-522309695>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAV3YM2ZZEBYWNKLNNQO3DQFEQF5ANCNFSM4H4MH5OQ>
> .
>

rogelio-o commented:
Understood! I've added a message handler with a context for those cases.

@tsegismont, If you have no time to review it before the release, let me know and I reverse the commits and make another PR with them.

rogelio-o commented:
Checks fail because a test not in the PR (WebClientTest.lambda$null$117:1351->lambda$testFileUploadsSameNameFormMultipart$113:1302->AsyncTestBase.assertEquals:360 expected:<test[2].txt> but was:<test[1].txt>).

PR 1345 is failing because of the same test.

tsegismont commented:
@rogelio-o thanks for all the added contribs. I'll have a look tonight and merge.

tsegismont commented:
@zenios thanks for reviewing!

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/ApolloWSHandlerImpl.java:
```diff
@@ -104,20 +113,25 @@ public void handle(RoutingContext routingContext) {
         }
 
-        switch (ApolloWSMessageType.from(type)) {
+        ApolloWSContext context = new ApolloWSContext(routingContext, serverWebSocket, type, message);
```
Do not create the object if the message handler is `null`

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/ApolloWSHandler.java:
```diff
@@ -73,4 +65,22 @@ static ApolloWSHandler create(GraphQL graphQL) {
   ApolloWSHandler dataLoaderRegistry(Function<RoutingContext, DataLoaderRegistry> factory);
 
+  /**
+   * Set interval seconds in which keepAlive messages are sent. If keepAlive is _null_ or
```
Please use `{@code null}` in Javadoc.

tsegismont replied:
Does it make sense to avoid completely keepalive messages? Shouldn't we use a sensible default instead?

zenios replied:
Default is 30s. It makes sense for somebody to disable it if it handles it externally

tsegismont replied:
In our case, how would users handle it externally?

rogelio-o replied:
With new changes, by default keep alive will be 30 and 0 value means keepalive disabled.

zenios replied:
Keep alive is just a message send from the server to the client every x seconds that triggers the client if not received to close the socket. The same can be done manually.

Example is someone already has a subscription in place that periodically sends some data and on the client if not received closes the socket

Also  on the Apollo implementation keep alive is optional and if not passed on construction it will not be enabled.

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/ApolloWSHandler.java:
```diff
@@ -73,4 +65,22 @@ static ApolloWSHandler create(GraphQL graphQL) {
+   */
+  @Fluent
+  @GenIgnore(GenIgnore.PERMITTED_TYPE)
+  ApolloWSHandler keepAlive(Long keepAlive);
```
Can you move this setting to an options object? It shouldn't change after initial setup.

rogelio-o commented:
> @rogelio-o a few comments/concerns inline.
> 
> Also, I don't believe we should use the routing context as input for query context and data loader initialization. Not sure what kind of objects it may retain. Instead, we should use the `ServerWebSocket`, and add a connection handler that allows the user to capture some information when the upgrade happens.
> 
> Then we need some doc about the new features.
> 
> Just tell me what you think and I can take of this tomorrow to make sure the PR lands on time for 3.8.1.

I've added the ApolloWSContext as argument of context and loaderRegistry functions.

I've also added the connection handler and fixes all the comments/concerns.

zenios commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/ApolloWSHandlerImpl.java:
```diff
@@ -115,6 +121,10 @@ public void handle(RoutingContext routingContext) {
   private void handleConnection(RoutingContext routingContext, ServerWebSocket serverWebSocket) {
     Map<String, Subscription> subscriptions = new ConcurrentHashMap<>();
 
+    if (connectionHandler != null) {
```
What do you think of merging connectionHandler and endHandler with messageHandler and add two more types (SOCKET_CREATED, SOCKET_DESTROYED) ?

rogelio-o replied:
I think connection and end handlers have different purpose than message handler. Connection and disconnection messages are handled by the message handler, but connection and disconnection “action” are handled by these handlers. The information we have is different.

tsegismont replied:
@zenios good catch

@rogelio-o please do not update the PR, I'm merging it. I will update this part of the code.

zenios commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/ApolloWSHandlerImpl.java:
```diff
@@ -0,0 +1,322 @@
+    if(
+      headers.contains(HttpHeaders.CONNECTION)
+        &&
+      HEADER_CONNECTION_UPGRADE_VALUE.equals(headers.get(HttpHeaders.CONNECTION).toLowerCase())
```
Protocol specified that on Connection header multiple values might exists along with upgrade. This should change to includes instead of equals

tsegismont replied:
@zenios good catch

@rogelio-o please do not update the PR, I'm merging it. I will update this part of the code.

tsegismont commented:
@rogelio-o it just hit `master` !!! :clap:

I added a few changes, in particular made message and message type public as they can be used in user code and translated to other languages.

@zenios thanks again for your input!

rogelio-o commented:
> @rogelio-o it just hit `master` !!! 👏
> 
> I added a few changes, in particular made message and message type public as they can be used in user code and translated to other languages.
> 
> @zenios thanks again for your input!

Great!! Thanks! And thanks @zenios for reviewing it.

zenios commented:
Great news.

I went through the changes and there are some things that are better to be
addressed before releasing.

Through the Apollo client you have the option to set some parameters on the
connection init that will be passed as payload. Those parameters are mostly
used for authentication.
Right now we can handle the connection_init message but even if we
terminate the socket from within the handler the system will still proceed
onto the switch statement and actually execute the connect function.

Second issue is that without a routing context there is no place to save
the authenticated user on connection_init and afterwards pick it up into
the queryContextFactory in order to forward it onto the subscription.

Looking forward to your comments.

On Tue, Aug 20, 2019 at 8:26 PM Rogelio Orts <notifications@github.com>
wrote:

> @rogelio-o <https://github.com/rogelio-o> it just hit master !!! 👏
>
> I added a few changes, in particular made message and message type public
> as they can be used in user code and translated to other languages.
>
> @zenios <https://github.com/zenios> thanks again for your input!
>
> Great!! Thanks! And thanks @zenios <https://github.com/zenios> for
> reviewing it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-web/pull/1325?email_source=notifications&email_token=AAAV3YICLCCOEHSUHRMBI63QFQSL5A5CNFSM4H4MH5O2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4XBQDQ#issuecomment-523114510>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAV3YOLZKW4BE3RGV5IY7TQFQSL5ANCNFSM4H4MH5OQ>
> .
>

tsegismont commented:
@zenios thanks for the comments, they are really appreciated!

With respect to authentication, I am fine if for now we only support basic/cookie auth. The browser will send headers that can be retrieved on the server with `ServerWebSocket#headers()`. To deny access, users can invoke `ServerWebSocket#setHandshake()` in a `connectionHandler` and fail the promise.

Once the user is authenticated, it can be stored in a map in verticle code with the `ServerWebSocket` as key.

Admittedly, vertx-web-graphql is not yet on a par with Apollo's server in terms of features. But this module is on tech preview and I would rather ship what we have than block the release.

However, please file an issue with your comments, we will address them in a later release.
